synapse/rest/client/room.py

```diff
@@ -587,14 +620,21 @@ class RoomMessageListRestServlet(RestServlet):
     def __init__(self, hs: "HomeServer"):
         super().__init__()
         self._hs = hs
+        self.clock = hs.get_clock()
         self.pagination_handler = hs.get_pagination_handler()
         self.auth = hs.get_auth()
         self.store = hs.get_datastores().main
 
-    @messsages_response_timer.time()
     async def on_GET(
         self, request: SynapseRequest, room_id: str
     ) -> Tuple[int, JsonDict]:
+        processing_start_time = self.clock.time_msec()
+        # Fire off and hope that we get a result by the end.
+        room_member_count_deferred = run_in_background(
+            self.store.get_number_joined_users_in_room,
+            room_id,
+        )
```

Comment:
Is `mypy` wrong here? The current setup gives a result at the end.

https://github.com/matrix-org/synapse/runs/7901466203?check_suite_focus=true
```
synapse/rest/client/room.py:633: error: Missing positional argument "room_id" in call to "run_in_background"  [call-arg]
synapse/rest/client/room.py:635: error: Argument 2 to "run_in_background" has incompatible type "str"; expected "RoomMemberWorkerStore"  [arg-type]
```

And if I add in the `self` parameter, it doesn't even work although it is talking about argument 2 and not 1 🤔:

```py
room_member_count_deferred = run_in_background(
    self.store.get_number_joined_users_in_room,
    self.store,
    room_id,
)
```

```
2022-08-18 09:58:45,970 - synapse.http.server - 185 - ERROR - GET-0 - Failed handle request via 'RoomMessageListRestServlet': <XForwardedForRequest at 0x106467d00 method='GET' uri='/_matrix/client/r0/rooms/!HBehERstyQBxyJDLfR%3Amy.synapse.server/messages?dir=b' clientproto='HTTP/1.1' site='8008'>
Traceback (most recent call last):
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 1660, in _inlineCallbacks
    result = current_context.run(gen.send, result)
StopIteration: {}

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/eric/Documents/github/element/synapse/.venv/lib/python3.9/site-packages/twisted/internet/defer.py", line 1660, in _inlineCallbacks
    result = current_context.run(gen.send, result)
StopIteration: {'$aaZt58h4vtfoFTXQTs6GtXpjV-iqdAPo1SC1XXiJta4': None, '$3xmrpHw9oe_BsjMVSwZOKCvK8o89CJAPA_N8Oh_qSOs': None, '$RoowBx-NKKs65f_FtuifEIYb5xZz12XYttIpRRVFw9w': None, '$0kQBsMp58Vqfoa-HfOp5-9jHh9CZ9OaJ4EB1ygWjeRA': None, '$BGbUG51rD3IYYDcg9Kka7cpydGE5LFO3BGg0mzXgWSE': None, '$2xC5ZzWvaGiHRFph1Smev61i3UStaeIT3FVn8oY38QA': None, '$Ufg0d0pwXCWid2ulx4FVrbG2IPcgMP2ML413G2_iAHM': None, '$kfuSEOfoHjRUVZrJczZkms8rvFmVBZCceu-4TMIfqZ8': None}

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/eric/Documents/github/element/synapse/synapse/http/server.py", line 368, in _async_render_wrapper
    callback_return = await self._async_render(request)
  File "/Users/eric/Documents/github/element/synapse/synapse/http/server.py", line 574, in _async_render
    callback_return = await raw_callback_return
  File "/Users/eric/Documents/github/element/synapse/synapse/rest/client/room.py", line 670, in on_GET
    room_member_count = await make_deferred_yieldable(room_member_count_deferred)
  File "/Users/eric/Documents/github/element/synapse/synapse/logging/context.py", line 804, in run_in_background
    res = f(*args, **kwargs)
TypeError: get_number_joined_users_in_room() takes 2 positional arguments but 3 were given
```

Reply:
I added these lines:

```python
        reveal_type(self.store.get_number_joined_users_in_room)
        reveal_type(room_id)
        reveal_type(run_in_background)
        room_member_count_deferred = run_in_background(
            self.store.get_number_joined_users_in_room,
            room_id,
        )
```

I got:
```
synapse/rest/client/room.py:633: note: Revealed type is "synapse.util.caches.descriptors._CachedFunction[def (self: synapse.storage.databases.main.roommember.RoomMemberWorkerStore, room_id: builtins.str) -> typing.Coroutine[Any, Any, builtins.int]]"
synapse/rest/client/room.py:634: note: Revealed type is "builtins.str"
synapse/rest/client/room.py:635: note: Revealed type is "Overload(def [P, R] (f: def (*P.args, **P.kwargs) -> typing.Awaitable[R`-2], *P.args, **P.kwargs) -> twisted.internet.defer.Deferred[R`-2], def [P, R] (f: def (*P.args, **P.kwargs) -> R`-2, *P.args, **P.kwargs) -> twisted.internet.defer.Deferred[R`-2])"
synapse/rest/client/room.py:636: error: Missing positional argument "room_id" in call to "run_in_background"  [call-arg]
synapse/rest/client/room.py:638: error: Argument 2 to "run_in_background" has incompatible type "str"; expected "RoomMemberWorkerStore"  [arg-type]
```

`_CachedFunction` raises a red flag here; our caching stuff plays badly with the type system at times I'm afraid.

Indeed, if I comment out `@cached()` on the function, I don't get the errors.
Mypy is misguided here, your code is fine — I guess we have to put a type ignore on it :-(
Might be able to do some ParamSpec wizardry with the cache functions (not sure how well support is nowadays... https://github.com/matrix-org/synapse/issues/11711).

Reply:
Thanks for the debugging steps here!

Added some `# type: ignore` with a comment why ✅